src/Core/GeometryPrimitives/Tensor.h

```diff
@@ -45,6 +45,7 @@
 #include <Core/Containers/Array1.h>
 #include <Core/GeometryPrimitives/Vector.h>
 #include <Core/GeometryPrimitives/share.h>
+#include <Core/Datatypes/DenseColumnMatrix.h>
```

Comment:
This creates a circular dependency of libraries. You'll need to implement the Tensor stuff that depends on DenseMatrix math in a third library, or return lower-level objects, like a `vector<double>`. Looking at the implementation of `mandel`, there is no Eigen usage, it's just a container you need. You can use Eigen like above, but only internally.

Reply:
In the module I'll be using it in, I'll be needing it stored as a Eigen vector to do an outer product to make it a 6x6 matrix. I was able to get it building by including <Eigen/Dense> directly. Is this bad practice?

Reply:
Including eigen directly is ok from a build perspective since it’s header only, like glm. But then intertwining type signatures with eigen’s types leads to the same problem I wrote that long post about related to returning glm types from renderer classes, find it in slack. It sounds like you need some intermediate types in a third library that can safely depend on both geometryprimitives and datatypes. Why not just make a new library called TensorVisualization, with classes for doing all these new things with tensors? That will keep the original tensor class from getting too bloated (like the field code), and the dependencies clearer.

Reply:
This mandel function only changes the notation. I'll move the functions(symmetricTensorFromNineElementArray, symmetricTensorFromMandel, symmetricTensorFromSixElementArray) that convert tensors to and from other types into another class called TensorConversion. With that, it'd still leave one usage of Eigen that I added in build_eigens_from_mat so I could use Eigen's eigen solver. The only other functions I added were a division operator and one to reorder tensor values.

Reply:
Wouldnt it be nice to start from scratch though? With a new tensor class you can do whatever you want with it without worrying about backward compatibility or dependency management. This is a more clear-cut case than the BBox addition. It will also greatly speed up your development if you don't have to recompile all the field code every time.

Reply:
Recompiling field classes doesn't take too long on 24 threads, but it'd be nice to make a tensor class from scratch. Eigen has a tensor class that's a wrapper around the matrix class. It's listed in their (unsupported modules)[https://eigen.tuxfamily.org/dox/unsupported/eigen_tensors.html] however. (This)[http://eigen.tuxfamily.org/index.php?title=Tensor_support] is a list of their features.

Reply:
One problem is that I do need the tensor fields to use the new eigen solver in Eigen, unless I wrote an eigen solver from scratch.

Reply:
OK how about this:
new class Tensor2/NewTensor/BetterTensor/AwesomeTensor/TarkTensor inherits from Eigen::Tensor<double,3>, lives in Core_Datatypes_Matrix to centralize the Eigen dependency.

We can add field support easily, that doesn't save the recompile cost though. Or we just add conversion functions between old and new, with care for the IO functions.

Reply:
Then old Tensor class is closed, nothing to add there; only add new stuff to new class.

Reply:
Let's name it SymmetricTensor since that's all we deal with in SCIRun. The documentation says they have optimizations for symmetry so that's a plus. I'd rather not do the conversion from old tensor to new tensor because it'll have to happen every time the new tensor is needed. I'd also have to convert back to make a field. What do you have in mind with the field support?

Reply:
We must have to add it as a field data type like the recent complex fields. But we might want to replace it, and come up with some code to handle the IO part separately.

Reply:
Are you suggesting that we have 2 different field types that are Tensor and NewTensor?